secretstores/tencent/ssm/ssm.go

```diff
@@ -0,0 +1,178 @@
+	ssm "github.com/tencentcloud/tencentcloud-sdk-go/tencentcloud/ssm/v20190923"
+)
+
+const (
+	SecretID         = "SecretID"
+	SecretKey        = "SecretKey"
+	Token            = "Token"
+	Region           = "Region"
+	VersionID        = "VersionID"
+	RequestID        = "RequestID"
+	pageLimit uint64 = 100
+)
+
+type ssmClient interface {
+	ListSecretsWithContext(ctx context.Context, request *ssm.ListSecretsRequest) (*ssm.ListSecretsResponse, error)
+	GetSecretValueWithContext(ctx context.Context, request *ssm.GetSecretValueRequest) (*ssm.GetSecretValueResponse, error)
+}
+
+type ssmSecretStore struct {
+	client ssmClient
+	logger logger.Logger
+}
+
+var (
+	_ secretstores.SecretStore = (*ssmSecretStore)(nil)
+)
+
+// NewSSM returns a new TencentCloud ssm secret store.
+func NewSSM(logger logger.Logger) secretstores.SecretStore {
+	return &ssmSecretStore{
+		logger: logger,
+	}
+}
+
+// Init creates a TencentCloud ssm client.
+func (s *ssmSecretStore) Init(metadata secretstores.Metadata) error {
+	var (
+		err    error
+		region string
+	)
+
+	secretID := metadata.Properties[SecretID]
+	secretKey := metadata.Properties[SecretKey]
+	token := metadata.Properties[Token]
+	region = metadata.Properties[Region]
+	if secretID == "" || secretKey == "" {
+		return errors.New("secret params are empty")
+	}
+
+	credential := common.NewTokenCredential(secretID, secretKey, token)
+	s.client, err = ssm.NewClient(credential, region, profile.NewClientProfile())
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// GetSecret retrieves a secret using a key and returns a map of decrypted string/string values.
+func (s *ssmSecretStore) GetSecret(ctx context.Context, req secretstores.GetSecretRequest) (secretstores.GetSecretResponse, error) {
+	response := secretstores.GetSecretResponse{
+		Data: map[string]string{},
+	}
+	if req.Name == "" {
+		return response, errors.New("secret name is empty")
+	}
+
+	versionID := req.Metadata[VersionID]
+	ssmReq := ssm.NewGetSecretValueRequest()
+	ssmReq.SecretName = &req.Name
+	ssmReq.VersionId = &versionID
+	ssResp, err := s.client.GetSecretValueWithContext(ctx, ssmReq)
+	if err != nil {
+		return response, err
+	}
+	var val string
+	ssResponse := ssResp.Response
+	if ssResponse.SecretBinary != nil {
+		bts, err := base64.StdEncoding.DecodeString(*ssResponse.SecretBinary)
+		if err != nil {
+			return response, err
+		}
+		val = string(bts)
+	}
+	if ssResponse.SecretString != nil {
+		val = *ssResponse.SecretString
+	}
+	response.Data[*ssResponse.SecretName] = val
+	response.Data[RequestID] = *ssResponse.RequestId
+
+	return response, nil
+}
+
+func (s *ssmSecretStore) BulkGetSecret(ctx context.Context, req secretstores.BulkGetSecretRequest) (secretstores.BulkGetSecretResponse, error) {
+	response := secretstores.BulkGetSecretResponse{
+		Data: map[string]map[string]string{},
+	}
+	versionID := req.Metadata[VersionID]
+
+	var start uint64 = 0
+	names, err := s.getSecretNames(ctx, &start)
```

Comment:
Shouldn't this read only the secret whose names is in the request object? Rather than _all_ secrets in the vault, which could be a lot?

Reply:
Instead of getting all the key-value APIs at once, we only get the key list through the `ListKeys` API first, and then traverse to get all the values.

Reply:
You're right, I misunderstood how the bulk get secrets API worked. This seems fine.

The problem is: what happens if the list of secrets changes between the time you invoke `ListKeys` and when you retrieve each secret?

Reply:
This second-level error, I think it should be acceptable.

Reply:
In fact, I think this batch access to key-values `BulkGetSecrets`API is a bit problematic. Because most secret cloud products need to specify `versionid`, in fact, this API cannot specify different versions, and users do not know which keys are there. So it feels a bit tasteless